src/common/libkvs/kvs_watch.h

```diff
@@ -5,45 +5,76 @@
                              int errnum);
 
-/* kvs_watch* is like kvs_get* except the registered callback is called
- * to set the value.  It will be called immediately to set the initial
- * value and again each time the value changes.
- * Any storage associated with the value given the
- * callback is freed when the callback returns.  If a value is unset, the
- * callback gets errnum = ENOENT.
+/* Register 'set' callback on non-directory 'key'.
+ * Callback is triggered once during registration to get the initial value.
+ * Once the reactor is (re-)entered, it will then be called each time the
+ * key changes.
  */
 int flux_kvs_watch (flux_t *h, const char *key, kvs_set_f set, void *arg);
+
+/* Register 'set' callback on directory 'key'.
+ * Callback is triggered once during registration to get the initial value,
+ * and thereafter, each time the directory changes. Note that due to the
+ * KVS's hash tree namespace organization, this function will be called
+ * whenever any key under this directory changes, since that forces the
+ * hash references to change on parents, all the way to the root.
+ */
 int flux_kvs_watch_dir (flux_t *h, kvs_set_dir_f set, void *arg,
                         const char *fmt, ...)
                         __attribute__ ((format (printf, 4, 5)));
 
-/* Cancel a kvs_watch, freeing server-side state, and unregistering any
- * callback.  Returns 0 on success, or -1 with errno set on error.
+/* Cancel a flux_kvs_watch(), freeing server-side state, and unregistering
+ * any callback.  Returns 0 on success, or -1 with errno set on error.
  */
 int flux_kvs_unwatch (flux_t *h, const char *key);
 
-/* While the above callback interface makes sense in plugin context,
- * the following is better for API context.  'json_str', 'dirp', and
- * 'valp' are IN/OUT parameters.  You should first read the current
- * value, then pass it into the respective kvs_watch_once call, which
- * will return with a new value when it changes.  (The original value
- * is freed inside the function; the new one must be freed by the
- * caller).  *json_str, *dirp, and *valp may be passed in with a NULL
- * value.  If the key is not set, ENOENT is returned without affecting
- * *valp.
- * FIXME: add more types.
+/* Block until 'key' changes from value represented by '*json_str'.
+ * 'json_str' is an IN/OUT parameter;  that is, it used to construct
+ * the watch RPC, then upon receipt of a watch response, it is freed
+ * and set to the new value.  Upon return, the caller should free
+ * the new value.
+ *
+ * 'json_str' may initially point to a NULL value.  The function will
+ * wait until 'key' exists then return its new value.
+ *
+ * If 'key' initially exists, then is removed, the function fails with
+ * ENOENT and the initial value is not freed.
  */
 int flux_kvs_watch_once (flux_t *h, const char *key, char **json_str);
+
+/* Same as above except value is a direcory pointed to by 'dirp'.
```

Comment:
typo ```directory```

Reply:
raced on comments/old commit and github is showing my comment as outdated.  typo ```directory```